ibacm/src/libacm.c

```diff
@@ -59,38 +59,37 @@ static void acm_set_server_port(void)
 int ib_acm_connect(char *dest)
 {
 	struct addrinfo hint, *res;
-	int ret;
+	int ret = 0;
 
 	acm_set_server_port();
 	memset(&hint, 0, sizeof hint);
 
 	if (dest && *dest != '/') {
 		hint.ai_family = AF_INET;
-		hint.ai_family = AF_UNSPEC;
+		hint.ai_protocol = IPPROTO_TCP;
 
 		ret = getaddrinfo(dest, NULL, &hint, &res);
-		if (ret)
-			return ret;
+		if (ret) {
+			if (ret == EAI_SYSTEM)
+				return errno;
+			else
+				return -1;
+		}
 
 		sock = socket(res->ai_family, res->ai_socktype,
 			      res->ai_protocol);
 		if (sock == -1) {
 			ret = errno;
-			goto err1;
+		} else {
+			((struct sockaddr_in *) res->ai_addr)->sin_port =
+				htobe16(server_port);
+			ret = connect(sock, res->ai_addr, res->ai_addrlen);
+			if (ret) {
+				ret = errno;
+				close(sock);
+				sock = -1;
+			}
 		}
-
-		((struct sockaddr_in *) res->ai_addr)->sin_port =
-			htobe16(server_port);
-		ret = connect(sock, res->ai_addr, res->ai_addrlen);
-		if (ret)
```

Comment:
I see a lot of clean up in this patch which I'm not sure is necessary.  Not wrong but...  For example AFAICT this hunk is only changing the return value to be errno.  The rest of the logic seems to be the same between the old and new code.  Furthermore, the new code does not follow the kernel convention of branch on failure.  So I'm not sure it is worth the change.  But since this is not the kernel I'm not going to argue against the change.  What I will argue for is a separate clean up patch which is separate from the logic change.

You also don't mention in the changelog that you are making the return meaning of ib_acm_connect consistent with this change and the one below.  That would help.

Reply:
There are four things being fixed by this patch:

1 - Removing the double setting of hint.ai_family and instead setting hint.ai_protocol.
2 - Making the return from the function consistent (i.e., return errno or -1) since the calling function, query_svcs(), assumed errno which was sometimes not the case.
3 - Fixed double freeaddinfo() in the server_mode != unix case successful return path.
4 - fixed the closing of the socket in the server_mode != unix successful return path.

Maybe this needed to be added to the Changelog?

The clean up of the patch, was mostly a result of fixing 3 and 4. Is it really preferred convention to use goto instead of if-else construct? I have to admit that I find that both surprising and ugly. It is true that I didn't need to touch the server-mode == unix case. However, I thought it best if the two cases were consistent. If that means using goto instead of if-else and (in my opinion) being less elegant, I can do that. 

I'll see if I can provide two patches. One that fixes the bugs (using the goto convention). And then (assuming that it seems worthwhile), one that structures the code using if-else.

Reply:
Sorry for the late response.

I think the fact that you list 4 things being fixed proves my point that there are too many things going on to properly review if this is correct.

github is confusing me right now.  Did you split the patch into 2?  and repush?  Ah I see it shows the new patch with a new version of this one...  Let me look at those.

Reply:
Before I go through a third iteration. Let me make sure I understand the requirements.

You only want to see one patch for each thing I'm fixing above? Seems excessive, but I can do that.

You (or rather, kernel convention) requires a goto or redundant code to bail out on error? That is don't use the if/else construct. That is counter to years of experience and education. But if that's the Linux way, I'll conform.

And yes, to split the patches I have created a new version of the original patch and then added the subsequent patch and the force pushed. If that's not the preferred method, please let me know what is.

Reply:
Yes.  But we are _not_ looking for redundant code.  The point is to not have redundant code if possible.  But also don’t put the “good” non-error code in an if/else.

https://www.kernel.org/doc/html/v4.10/process/coding-style.html#centralized-exiting-of-functions

We may be hitting an issue with this convention as well.

https://www.kernel.org/doc/html/v4.10/process/coding-style.html#functions

Is ib_acm_connect() trying to do too many things?

Reply:
It handles two cases. Is that too many things?